de.fu_berlin.inf.dpp.whiteboard/src/de/fu_berlin/inf/dpp/whiteboard/ui/browser/BrowserActionExecuter.java

```diff
@@ -0,0 +1,101 @@
+        .getLogger(BrowserActionExecuter.class);
+
+    private final SXEController sxe;
+
+    public BrowserActionExecuter(SXEController sxe) {
+        this.sxe = sxe;
+    }
+
+    public void execute(BrowserAction action) {
+        sxe.executeAndCommit(getChangeRecords(action));
+    }
+
+    private List<IRecord> getChangeRecords(BrowserAction action) {
+        String type = action.getType();
+        if (type.equals(RecordType.NEW.toString())) {
+            return create(action);
+        } else if (type.equals(RecordType.SET.toString())) {
+            return set(action);
+        } else if (type.equals(RecordType.REMOVE.toString())) {
+            return remove(action);
+        } else {
+            LOG.error("undefined action type received: " + type);
+            return Collections.emptyList();
+        }
+    }
+
+    private List<IRecord> create(BrowserAction action) {
+        List<IRecord> records = new ArrayList<IRecord>();
+        // since the library fabric.js (used in the html whiteboard) does not
+        // support parent hierarchies this function will always add the new
+        // element to the root element
+        ElementRecord parent = sxe.getDocumentRecord().getRoot();
+        ElementRecord newElement = parent.createNewElementRecord(null,
+            action.getElementType());
+        newElement.setRid(action.getID());
+        records.add(newElement);
+        records.addAll(setAttributes(newElement, action.getAttributes()));
+        return records;
+    }
+
+    private List<IRecord> setAttributes(ElementRecord element,
+        List<Map<String, String>> attributes) {
+        List<IRecord> records = new ArrayList<IRecord>();
+        for (Map<String, String> attribute : attributes) {
+            IRecord record = element.createNewOrSetAttributeRecord(null,
+                attribute.get("name"), attribute.get("chdata"), false);
+            if (record instanceof AttributeRecord)
+                ((AttributeRecord) record).setRid(attribute.get("rid"));
+            records.add(record);
+        }
+        return records;
+    }
+
+    private List<IRecord> set(BrowserAction action) {
+        ElementRecord element = getElement(action.getID());
+        if (element == null)
+            return Collections.emptyList();
+        return setAttributes(element, action.getAttributes());
+
+    }
+
+    private List<IRecord> remove(BrowserAction action) {
+        ElementRecord element = getElement(action.getID());
+        if (element == null)
+            return Collections.emptyList();
+
+        List<IRecord> records = new ArrayList<IRecord>();
+        records.add(element.getRemoveRecord());
+        return records;
+    }
+
+    private ElementRecord getElement(String id) {
+        try {
+            return sxe.getDocumentRecord().getElementRecordById(id);
+        } catch (MissingRecordException e) {
+            LOG.error("trying to set access an element that does NOT"
```

Comment:
-set

and why do you not add the id ? Currently this message is not that helpful if you want to debug the issue as the id is silently swallowed. Or is the id part already contained in the MissingRecordException ?

Reply:
the id does not help at all, it is only used internally to distinguish between the different elements and to quickly access and modify an element, but the actual id itself is just a random string:

https://github.com/saros-project/saros/blob/8d91a46607901e2cfcb2b7a7f19581d01705280e/de.fu_berlin.inf.dpp.whiteboard/src/de/fu_berlin/inf/dpp/whiteboard/sxe/records/NodeRecord.java#L94-L109